### DISCUSSION TITLE
Reasons for switching from `phpseclib` to `ssh` command

### DISCUSSION THREAD

**Discussion by binaryfire** [deployphp/deployer#3954] [General]
Hi guys

I noticed deployer now uses the `ssh` command under the hood instead of `phpseclib` (and it seems like [`phpseclib` was used in the past](https://github.com/deployphp/deployer/pull/712)).

I'm working on an app which needs to SSH into multiple servers and would love to get your feedback on why you made the change. Did you run into performance / scalability / concurrency issues with `phpseclib`?

Thanks!

**Comment by antonmedv** [comment]
Yes, we did indeed switch from phpseclib to the native SSH command (OpenSSH) for several reasons. OpenSSH offers improved performance and reliability, particularly for handling multiple simultaneous connections, which is essential for Deployer’s scaling needs.

One of the biggest advantages of using OpenSSH is its compatibility with newer SSH features and protocols, such as more modern encryption algorithms and enhanced key handling. These improvements help keep Deployer up-to-date with the latest security standards. OpenSSH also tends to have fewer bugs and edge cases in comparison to phpseclib, especially when working with various server configurations and network conditions.

The change to OpenSSH has helped streamline the deployment process and improve overall reliability, so it's a move we'd definitely recommend if your app requires robust, scalable SSH interactions.

**Reply by binaryfire** [reply]
@antonmedv Thanks, this is really helpful information. Much appreciated!
